tools/cldr-code/src/main/java/org/unicode/cldr/util/GlossonymConstructor.java

```diff
@@ -0,0 +1,124 @@
+     * It is non-null to satisfy TestPathHeadersAndValues.
+     * It should not be treated as an actual path; for example, the
+     * Survey Tool Info Panel should not show a broken "Jump to original" link.
+     */
+    public static final String PSEUDO_PATH = "constructed";
+
+    /**
+     * Is the given path eligible for getting a constructed value?
+     *
+     * @param xpath the given path
+     * @return true if eligible
+     */
+    public static boolean pathIsEligible(String xpath) {
+        return xpath.startsWith(PATH_PREFIX) && xpath.contains(CODE_SEPARATOR);
+    }
+
+    /**
+     * Is the given value bogus, and therefore eligible for getting replaced by a constructed value?
+     *
+     * @param value the given value
+     * @return true if bogus
+     */
+    public static boolean valueIsBogus(String value) {
+        return value == null || value.contains(CODE_SEPARATOR) || CheckNames.matchDigitPattern(value);
+    }
+
+    private final CLDRFile cldrFile;
+
+    public GlossonymConstructor(CLDRFile cldrFile) {
+        this.cldrFile = cldrFile;
+        if (!cldrFile.isResolved()) {
+            throw new IllegalArgumentException("Unresolved CLDRFile in GlossonymConstructor constructor");
+        }
+    }
+
+    /**
+     * Get the constructed value and fill in tracking information about where it was found
+     *
+     * @param xpath the path
+     * @param pathWhereFound if not null, to be filled in
+     * @param localeWhereFound if not null, to be filled in
+     * @return the constructed value, or null
+     */
+    public String getValueAndTrack(String xpath, Output<String> pathWhereFound, Output<String> localeWhereFound) {
+        final String constructedValue = getValue(xpath);
+        if (constructedValue != null) {
+            track(pathWhereFound, localeWhereFound);
+            return constructedValue;
+        }
+        return null;
+    }
+
+    /**
+     * Get the constructed value for the given path
+     *
+     * @param xpath the given path
+     * @return the constructed value, or null
+     */
+    public String getValue(String xpath) {
+        if (pathIsEligible(xpath)) {
+            return reallyGetValue(xpath);
+        }
+        return null;
+    }
+
+    private synchronized String reallyGetValue(String xpath) {
+        final XPathParts parts = XPathParts.getFrozenInstance(xpath);
+        final String type = parts.getAttributeValue(-1, "type");
+        if (type.contains(CODE_SEPARATOR)) {
+            final String alt = parts.getAttributeValue(-1, "alt");
+            final CLDRFile.SimpleAltPicker altPicker = (alt == null) ? null : new CLDRFile.SimpleAltPicker(alt);
+            final String value = cldrFile.getName(type, true, altPicker);
+            if (!valueIsBogus(value)) {
+                return value;
+            }
+        }
+        return null;
+    }
+
+    private void track(Output<String> pathWhereFound, Output<String> localeWhereFound) {
```

Comment:
In some cases it is harder to read to have extremely small private methods, because you have to look somewhere else to figure out what is actually happening. I would inline this method, for example.

Reply:
OK, I'll inline it. Part of my motivation for separating it was anticipation of an improved tracking method with its own class along the lines of XMLSource.AliasLocation but more general.